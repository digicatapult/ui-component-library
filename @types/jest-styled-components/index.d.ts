declare module 'jest-styled-components/serializer'
