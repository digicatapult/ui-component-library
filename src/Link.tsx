import styled from 'styled-components'
import {system} from 'styled-system'
import {get} from './constants'
import sx, {SxProp} from './sx'
import {ComponentProps} from './utils/types'

type StyledLinkProps = {
  hoverColor?: string
  muted?: boolean
  underline?: boolean
} & SxProp

const hoverColor = system({
  hoverColor: {
    property: 'color',
    scale: 'colors',
  },
})

const Link = styled.a<StyledLinkProps>`
  color: ${props => (props.muted ? get('colors.fg.muted')(props) : get('colors.accent.fg')(props))};
  text-decoration: ${props => (props.underline ? 'underline' : 'none')};
  &:hover {
    text-decoration: ${props => (props.muted ? 'none' : 'underline')};
    ${props => (props.hoverColor ? hoverColor : props.muted ? `color: ${get('colors.accent.fg')(props)}` : '')};
  }
  &:is(button) {
    display: inline-block;
    padding: 0;
    font-size: inherit;
    white-space: nowrap;
    cursor: pointer;
    user-select: none;
    background-color: transparent;
    border: 0;
    appearance: none;
  }
  ${sx};
`

export type LinkProps = ComponentProps<typeof Link>
export default Link
