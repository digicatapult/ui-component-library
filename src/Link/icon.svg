<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 20010904//EN"
 "http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd">
<svg version="1.0" xmlns="http://www.w3.org/2000/svg"
 width="200.000000pt" height="200.000000pt" viewBox="0 0 200.000000 200.000000"
 preserveAspectRatio="xMidYMid meet">

<g transform="translate(0.000000,200.000000) scale(0.100000,-0.100000)"
fill="#000000" stroke="none">
<path d="M1192 1807 c-30 -32 -29 -87 1 -115 22 -21 33 -22 185 -22 89 0 162
-3 162 -8 0 -4 -196 -203 -435 -442 -411 -411 -435 -437 -435 -471 0 -46 33
-79 81 -79 32 0 64 30 469 435 239 239 438 435 442 435 5 0 8 -73 8 -163 0
-158 1 -164 23 -185 32 -30 87 -29 115 1 22 23 22 30 22 306 l0 282 -25 24
-24 25 -284 0 c-283 0 -284 0 -305 -23z"/>
<path d="M315 1561 c-22 -10 -57 -36 -77 -57 -70 -73 -68 -55 -68 -631 0 -582
-2 -560 76 -635 73 -70 55 -68 631 -68 582 0 560 -2 635 76 67 71 68 74 68
453 0 187 -3 347 -6 356 -10 25 -62 47 -94 40 -55 -12 -55 -11 -60 -376 -5
-336 -5 -339 -28 -361 l-23 -23 -494 0 -494 0 -23 23 -23 23 0 494 0 494 23
23 c22 23 25 23 365 28 188 3 343 7 344 8 1 1 10 16 19 33 18 36 13 70 -16 99
-19 19 -35 20 -367 20 -318 0 -351 -2 -388 -19z"/>
</g>
</svg>
