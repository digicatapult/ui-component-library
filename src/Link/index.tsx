import React from 'react'
import styled from 'styled-components'
const iconSvg = require('./icon.svg') as string

export interface LinkProps {
  text: string
  href: string
  openInNewTab?: boolean
  color?: string
  visitedColor?: string
  iconSrc?: string
  iconHeight?: string
  iconMargin?: string
}

export interface AProps {
  visitedColor: string
}

const A = styled.a<AProps>`
  color: ${({ color }) => color};

  :visited {
    color: ${({ visitedColor }) => visitedColor};
  }
`

export interface IconProps {
  height: string
  margin: string
}

const Icon = styled.img<IconProps>`
  height: ${({ height }) => height};
  margin: ${({ margin }) => margin};
  vertical-align: top;
`

const Link: React.FC<LinkProps> = ({
  text,
  href,
  openInNewTab = true,
  color = 'black',
  visitedColor = 'black',
  iconSrc = iconSvg,
  iconHeight = '1em',
  iconMargin = '0 0.5em',
}) => {
  const target = openInNewTab ? '_blank' : '_self'

  return (
    <span>
      <A href={href} target={target} color={color} visitedColor={visitedColor}>
        {text}
      </A>
      <Icon
        src={iconSrc}
        alt="link icon"
        height={iconHeight}
        margin={iconMargin}
      />
    </span>
  )
}

export default Link
