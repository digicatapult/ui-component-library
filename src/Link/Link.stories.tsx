import React from 'react'
import { Story, StoryObj } from '@storybook/react'

import Link from './index.js'

export default {
  title: 'Components/Link',
  component: Link,
  argTypes: {
    text: {
      control: 'text',
    },
    href: {
      control: 'text',
    },
    openInNewTab: {
      control: 'boolean',
    },
    color: {
      control: 'color',
    },
    visitedColor: {
      control: 'color',
    },
    iconHeight: {
      control: 'text',
    },
    iconMargin: {
      control: 'text',
    },
  },
}

export const Default: StoryObj = {
  args: {
    text: 'Digital Catapult',
    href: 'https://www.digicatapult.org.uk/',
  },
}
export const HII: StoryObj = {
  args: {
    text: 'Link to project',
    href: 'https://gtr.ukri.org/projects?ref=EP%2FG06279X%2F1',
  },
}
