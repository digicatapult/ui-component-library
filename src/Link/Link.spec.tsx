import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Link from './index.js'

describe('Link', () => {
  test('default', () => {
    const tree = renderer.create(<Link text="test link" href="/" />).toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('overrides', () => {
    const tree = renderer
      .create(
        <Link
          text="test link"
          href="/"
          openInNewTab={false}
          color="blue"
          visitedColor="purple"
          iconHeight="0.5em"
          iconMargin="0"
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
