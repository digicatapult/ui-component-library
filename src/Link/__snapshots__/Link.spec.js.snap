// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link default 1`] = `
.c0 {
  color: black;
}

.c0:visited {
  color: black;
}

.c1 {
  height: 1em;
  margin: 0 0.5em;
  vertical-align: top;
}

<span>
  <a
    className="c0"
    color="black"
    href="/"
    target="_blank"
  >
    test link
  </a>
  <img
    alt="link icon"
    className="c1"
    height="1em"
    src="/images/link.png"
  />
</span>
`;

exports[`Link overrides 1`] = `
.c0 {
  color: blue;
}

.c0:visited {
  color: purple;
}

.c1 {
  height: 0.5em;
  margin: 0;
  vertical-align: top;
}

<span>
  <a
    className="c0"
    color="blue"
    href="/"
    target="_self"
  >
    test link
  </a>
  <img
    alt="link icon"
    className="c1"
    height="0.5em"
    src="/images/link.png"
  />
</span>
`;
